httpcore5/src/main/java/org/apache/hc/core5/http/protocol/ForwardedRequest.java

```diff
@@ -0,0 +1,159 @@
+ * <p>
+ * The Forwarded header is added to the request by calling the `HttpRequest.addHeader` method. If a
+ * Forwarded header already exists in the request, the new key-value pairs are appended to the
+ * existing header.
+ *
+ * @since 5.3
+ */
+@Contract(threading = ThreadingBehavior.IMMUTABLE)
+public class ForwardedRequest implements HttpRequestInterceptor {
+
+    /**
+     * The name of the header to set in the HTTP request.
+     */
+    private static final String X_FORWARDED_HEADER_NAME = "X-Forwarded-For";
+
+    /**
+     * The name of the header to set in the HTTP request.
+     */
+    private static final String FORWARDED_HEADER_NAME = "Forwarded";
+
+
+    /**
+     * Singleton instance.
+     */
+    public static final HttpRequestInterceptor INSTANCE = new ForwardedRequest();
+
+
+    /**
+     * Adds a Forwarded header to the specified HTTP request.
+     *
+     * @param request the HTTP request to add the Forwarded header to
+     * @param entity  the entity details associated with the request, or null if the request does
+     *                not contain an entity
+     * @param context the HTTP context for the request
+     * @throws HttpException if an error occurs while adding the header
+     * @throws IOException   if an I/O error occurs while adding the header
+     */
+    @Override
+    public void process(final HttpRequest request, final EntityDetails entity, final HttpContext context) throws HttpException, IOException {
+
+        Args.notNull(request, "HTTP request");
+        Args.notNull(context, "HTTP context");
+
+        final ProtocolVersion ver = context.getProtocolVersion() != null ? context.getProtocolVersion() : HttpVersion.HTTP_1_1;
+
+        final URIAuthority authority = request.getAuthority();
+        if (authority == null) {
+            throw new ProtocolException("Request authority not specified");
+        }
+
+        final int port = authority.getPort();
+
+        final StringBuilder valueBuilder = new StringBuilder();
+        valueBuilder.append("host=").append(authority.getHostName());
+
+        if (port != -1) {
+            valueBuilder.append(";port=").append(port);
+        }
+
+        final String protoValue = ver.getProtocol();
+        if (protoValue != null) {
+            valueBuilder.append("proto=").append(protoValue);
+        }
+
+
+        final EndpointDetails endpointDetails = (EndpointDetails) context.getAttribute(HttpCoreContext.CONNECTION_ENDPOINT);
+
+        // Add the "for" property
+        if (endpointDetails != null) {
+            valueBuilder.append(";for=").append(endpointDetails.getRemoteAddress().toString());
+        }
+
+        final String byValue = (String) context.getAttribute(X_FORWARDED_HEADER_NAME);
+        if (byValue != null) {
+            valueBuilder.append(";by=").append(byValue);
+        } else {
+            valueBuilder.append(";by=unknown");
+        }
+
+        final Header header = request.getFirstHeader(X_FORWARDED_HEADER_NAME);
```

Comment:
@arturobernalg If all you want is to override a header value if it has already been set, use `Header#setHeader`. It is massively more efficient then `Header#getFirstHeader`, `Header#removeHeader`, `Header#addHeader` sequence.

Reply:
change it.

Reply:
@arturobernalg Why do you still need `Header#getFirstHeader` and `Header#removeHeader`?